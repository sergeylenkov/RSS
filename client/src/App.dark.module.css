.container {
    background-color: #212121;
}

.header {
    background: #1b1c1d;
}

.feeds {
    background-color: #212121;
}