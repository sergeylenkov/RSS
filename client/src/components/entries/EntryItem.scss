.entry-item {
	position: relative;
	flex-shrink: 0;
	width: 100%;
  margin: 0 0 20px 0;
  padding: 20px;
  box-sizing: border-box;
	background: #ffffff;

	&__title {
		margin: 0 0 15px 0;
		font-size: 18px;
		font-weight: bold;
		line-height: 22px;
	}

	&__title a {
		color: #000000;
		text-decoration: none;
	}

	&__description {
		position: relative;
		margin: 20px 0 0 0;
		font-size: 14px;
		line-height: 18px;
		color: #000000;
		overflow: hidden;
	}

	&_collapsed &__description {
		max-height: 800px;
	}

	&__description img {
		max-width: 100%;
		height: auto;
		margin: 0 auto 10px auto;
		padding: 0px;
	}

	&__description font {
		color: #000000 !important;
	}

	&__description p {
		margin: 0 0 10px 0;
	}

	&__description code {
		font-family: inherit;
	}

	&__description blockquote {
		font-style: italic;
		color: #767676;
		margin: 0;
	}

	&__description h1,
	&__description h2 {
		font-size: 14px;
			margin: 0;
	}

	&__description ol {
		margin: 0;
		padding: 0 0 0 20px;
	}

	&__description a {
		color: #4076cc;
	}

	&__description pre {
		white-space: pre-wrap;
	}
}

.entry-item-expand-button {
	position: absolute;
	left: 0;
	right: 0;
	bottom: 58px;
	height: 60px;
	width: 100%;
	background: rgba(255, 255, 255, 0.8);
	cursor: pointer;
}

.entry-item-feed {
	display: flex;
	margin: 10px 0;

	&__title {
		margin: auto 0 auto 10px;
		font-size: 12px;
		color: #909090;
	}

	&__icon {
		width: 14px;
		height: 14px;
		margin: auto 0;
		background-image: "";
		background-repeat: no-repeat;
		background-size: cover;
		background-color: #ebebeb;
	}
}

.entry-item-info {
	display: flex;
	width: 100%;
	margin: 20px 0 0 0;

	&__item {
		display: flex;
		margin: auto 15px auto 0;
		cursor: pointer;
	}

	&__icon {
		width: 18px;
		height: 18px;
		margin: auto 0;
		fill: #818181;
	}
}

/*
* theme
*/

.dark .entry {
	background-color: #1d1d1d;

	&__title a {
		color: #bfbfbf;
	}

	&__description {
		color: #bfbfbf;
	}

	&__description font {
		color: #bfbfbf !important;
	}

	&__expandButton {
		background-color: rgba(29, 29, 29, 0.8);
	}
}