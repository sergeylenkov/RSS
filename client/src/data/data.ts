export interface SetReadResponse {
  id: number;
  isRead: boolean;
}

export interface SetFavoriteResponse {
  id: number;
  isFavorite: boolean;
}

export interface SetViewedResponse {
  ids: number[];
  isViewed: boolean;
}

export interface ClearEntriesResponse {
  days: number;
}

export interface UpdateFeedResponse {
  id: number;
  data: any;
}

export interface Feed {
  id: number;
  title: string;
  link: string;
  image: string;
  icon: string;
  count: number;
}

export interface Entry {
  id: number;
  feedId: number;
  feed: Feed;
  title: string;
  description: string;
  link: string;
  isViewed: boolean;
  isFavorite: boolean;
  isRead: boolean;
}

export class Data {
  private url = '';
  private feeds: Feed[] = [];
  private feedsDict: { [key: string]: Feed } = {};

  constructor(url: string) {
    this.url = url;
  }

  public update() {
    return new Promise<any[]>((resolve, reject) => {
      fetch(`${this.url}feeds/update`).then((response) => {
        return response.json();
      }).then((data) => {
        this.updateFeedsInEntries(data);
        resolve(data);
      }).catch((error) => {
        reject(error);
      })
    });
  }

  public getFeeds() {
    return new Promise<Feed[]>((resolve) => {
      fetch(`${this.url}feeds`).then((response) => {
        return response.json();
      }).then((data: Feed[]) => {
        this.feeds = data;

        this.feeds.forEach((feed: Feed) => {
          const a = document.createElement('a');
          a.href = feed.link;

          if (feed.image && feed.image.length > 0) {
            feed.icon = feed.image;
          } else {
            feed.icon = `${a.protocol}//${a.hostname}/favicon.ico`;
          }

          this.feedsDict[feed.id] = feed;
        });

        resolve(this.feeds);
      });
    });
  }

  public allEntries() {
    return new Promise<Entry[]>((resolve) => {
      fetch(`${this.url}entries`).then((response) => {
        return response.json();
      }).then((data: Entry[]) => {
        this.updateFeedsInEntries(data);
        resolve(data);
      });
    });
  }

  public readEntries() {
    return new Promise<Entry[]>((resolve) => {
      fetch(`${this.url}entries/read`).then((response) => {
        return response.json();
      }).then((data: Entry[]) => {
        this.updateFeedsInEntries(data);
        resolve(data);
      });
    });
  }

  public getFavorites() {
    return new Promise<Entry[]>((resolve) => {
      fetch(`${this.url}entries/favorites`).then((response) => {
        return response.json();
      }).then((data: Entry[]) => {
        this.updateFeedsInEntries(data);
        resolve(data);
      });
    });
  }

  public getUnviewed() {
    return new Promise<Entry[]>((resolve) => {
      fetch(`${this.url}entries/unviewed`).then((response) => {
        return response.json();
      }).then((data: Entry[]) => {
        this.updateFeedsInEntries(data);
        resolve(data);
      });
    });
  }

  public setViewed(ids: number[]) {
    return new Promise<SetViewedResponse>((resolve) => {
      fetch(`${this.url}entries/view`, {
        method: 'POST',
        body: JSON.stringify({ ids: ids }),
        headers: {
          'Content-Type': 'application/json'
        }
      }).then((response) => {
        return response.json();
      }).then((data) => {
        resolve(data);
      });
    });
  }

  public setRead(id: number, isRead: boolean) {
    return new Promise<SetReadResponse>((resolve) => {
      const method = isRead ? 'PUT' : 'DELETE';

      fetch(`${this.url}entries/${id}/read`, {
        method: method
      }).then((response) => {
        return response.json();
      }).then((data) => {
        resolve(data);
      });
    });
  }

  public setFavorite(id: number, isFavorite: boolean) {
    return new Promise<SetFavoriteResponse>((resolve) => {
      const method = isFavorite ? 'PUT' : 'DELETE';

      fetch(`${this.url}entries/${id}/favorite`, {
        method: method
      }).then((response) => {
        return response.json();
      }).then((data) => {
        resolve(data);
      });
    });
  }

  public addFeed(link: string) {
    return new Promise((resolve) => {
      fetch(`${this.url}feeds`, {
        method: 'POST',
        body: JSON.stringify({ link: link }),
        headers: {
          'Content-Type': 'application/json'
        }
      }).then((response) => {
        return response.json();
      }).then((data) => {
        resolve(data);
      });
    });
  }

  public updateFeed(id: number, data: any) {
    return new Promise<UpdateFeedResponse>((resolve) => {
      fetch(`${this.url}feeds/${id}`, {
        method: 'PATCH',
        body: JSON.stringify(data),
        headers: {
          'Content-Type': 'application/json'
        }
      }).then((response) => {
        return response.json();
      }).then((data) => {
        resolve(data);
      });
    });
  }

  public deleteFeed(id: number) {
    return new Promise((resolve) => {
      fetch(`${this.url}feeds/${id}`, {
        method: 'DELETE'
      }).then((response) => {
        return response.json();
      }).then((data) => {
        resolve(data);
      });
    });
  }

  public clearEntries(days: number) {
    return new Promise<ClearEntriesResponse>((resolve) => {
      fetch(`${this.url}entries/clear/${days}`).then((response) => {
        return response.json();
      }).then((data) => {
        resolve(data);
      });
    });
  }

  public getFeedById(id: number) : Feed {
    return this.feedsDict[id];
  }

  public updateFeedsInEntries(entries: Entry[]) {
    entries.forEach(entry => {
      entry.feed = this.getFeedById(entry.feedId);
    });
  }
}
