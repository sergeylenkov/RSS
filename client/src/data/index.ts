import { config } from '../constants';

export interface SetReadResponse {
  id: number;
  isRead: boolean;
}

export interface SetFavoriteResponse {
  id: number;
  isFavorite: boolean;
}

export interface SetViewedResponse {
  ids: number[];
  isViewed: boolean;
}

export interface ClearEntriesResponse {
  days: number;
}

export interface UpdateFeedResponse {
  id: number;
  data: Feed;
}

export interface DeleteFeedResponse {
  id: number;
  isDeleted: boolean;
}

export interface Feed {
  id: number;
  title: string;
  link: string;
  image: string;
  icon: string;
  count: number;
}

export interface Entry {
  id: number;
  feedId: number;
  feed: Feed;
  title: string;
  description: string;
  link: string;
  isViewed: boolean;
  isFavorite: boolean;
  isRead: boolean;
}

class Data {
  private _url: string;

  constructor(url: string) {
    this._url = url;
  }

  update() {
    return new Promise<Entry[]>((resolve, reject) => {
      fetch(`${this._url}feeds/update`).then((response) => {
        return response.json();
      }).then((data) => {
        resolve(data);
      }).catch((error) => {
        reject(error);
      })
    });
  }

  getFeeds() {
    return new Promise<Feed[]>((resolve) => {
      fetch(`${this._url}feeds`).then((response) => {
        return response.json();
      }).then((feeds: Feed[]) => {
        feeds.forEach((feed: Feed) => {
          const a = document.createElement('a');
          a.href = feed.link;

          if (feed.image && feed.image.length > 0) {
            feed.icon = feed.image;
          } else {
            feed.icon = `${a.protocol}//${a.hostname}/favicon.ico`;
          }
        });

        resolve(feeds);
      });
    });
  }

  allEntries() {
    return new Promise<Entry[]>((resolve) => {
      fetch(`${this._url}entries`).then((response) => {
        return response.json();
      }).then((entries: Entry[]) => {
        resolve(entries);
      });
    });
  }

  readEntries() {
    return new Promise<Entry[]>((resolve) => {
      fetch(`${this._url}entries/read`).then((response) => {
        return response.json();
      }).then((entries: Entry[]) => {
        resolve(entries);
      });
    });
  }

  getFavorites() {
    return new Promise<Entry[]>((resolve) => {
      fetch(`${this._url}entries/favorites`).then((response) => {
        return response.json();
      }).then((entries: Entry[]) => {
        resolve(entries);
      });
    });
  }

  getUnviewed() {
    return new Promise<Entry[]>((resolve) => {
      fetch(`${this._url}entries/unviewed`).then((response) => {
        return response.json();
      }).then((entries: Entry[]) => {
        resolve(entries);
      });
    });
  }

  setViewed(ids: number[]) {
    return new Promise<SetViewedResponse>((resolve) => {
      fetch(`${this._url}entries/view`, {
        method: 'POST',
        body: JSON.stringify({ ids: ids }),
        headers: {
          'Content-Type': 'application/json'
        }
      }).then((response) => {
        return response.json();
      }).then((response: SetViewedResponse) => {
        resolve(response);
      });
    });
  }

  setRead(id: number, isRead: boolean) {
    return new Promise<SetReadResponse>((resolve) => {
      const method = isRead ? 'PUT' : 'DELETE';

      fetch(`${this._url}entries/${id}/read`, {
        method: method
      }).then((response) => {
        return response.json();
      }).then((response: SetReadResponse) => {
        resolve(response);
      });
    });
  }

  setFavorite(id: number, isFavorite: boolean) {
    return new Promise<SetFavoriteResponse>((resolve) => {
      const method = isFavorite ? 'PUT' : 'DELETE';

      fetch(`${this._url}entries/${id}/favorite`, {
        method: method
      }).then((response) => {
        return response.json();
      }).then((response: SetFavoriteResponse) => {
        resolve(response);
      });
    });
  }

  addFeed(link: string) {
    return new Promise<Feed>((resolve) => {
      fetch(`${this._url}feeds`, {
        method: 'POST',
        body: JSON.stringify({ link: link }),
        headers: {
          'Content-Type': 'application/json'
        }
      }).then((response) => {
        return response.json();
      }).then((feed: Feed) => {
        resolve(feed);
      });
    });
  }

  updateFeed(id: number, data: string) {
    return new Promise<UpdateFeedResponse>((resolve) => {
      fetch(`${this._url}feeds/${id}`, {
        method: 'PATCH',
        body: JSON.stringify(data),
        headers: {
          'Content-Type': 'application/json'
        }
      }).then((response) => {
        return response.json();
      }).then((response: UpdateFeedResponse) => {
        resolve(response);
      });
    });
  }

  deleteFeed(id: number) {
    return new Promise<DeleteFeedResponse>((resolve) => {
      fetch(`${this._url}feeds/${id}`, {
        method: 'DELETE'
      }).then((response) => {
        return response.json();
      }).then((response: DeleteFeedResponse) => {
        resolve(response);
      });
    });
  }

  clearEntries(days: number) {
    return new Promise<ClearEntriesResponse>((resolve) => {
      fetch(`${this._url}entries/clear/${days}`).then((response) => {
        return response.json();
      }).then((response: ClearEntriesResponse) => {
        resolve(response);
      });
    });
  }
}

export default new Data(config.urls.api);