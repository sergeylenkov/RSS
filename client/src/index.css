body {
  margin: 0;
  padding: 0;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;  
}

button {
  border: none;
  background: none;
  outline: 0;
  padding: 0;
  margin: 0;
}