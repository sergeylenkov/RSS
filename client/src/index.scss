html {
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
  font-family: i8sans, 'Helvetica Neue', 'Arial', sans-serif;
}

body {
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
  background: #f0f0f0;
}

button {
  margin: 0;
  padding: 0;
  background: none;
  border: 0;
  outline: 0;
}

#root {
  width: 100%;
  height: 100%;
}

@keyframes spin {
  100% {
    transform: rotate(360deg);
  }
}

.fade-enter {
  opacity: 0;
}

.fade-enter-active {
  opacity: 1;
  transition: opacity 200ms ease-in;
}

.fade-exit {
  opacity: 1;
}

.fade-exit-active {
  opacity: 0;
  transition: opacity 200ms ease-out;
}
