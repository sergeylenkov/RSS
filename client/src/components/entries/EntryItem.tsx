import React, { useEffect, useRef, useState } from 'react';
import { FavoriteIcon, FavoriteSelectedIcon, ReadIcon } from '../Icons';
import { Entry } from '../../data';
import { debounce } from '../../utils';
import { isLong, removeSelfLinks } from '../../utils/entry';
import { Bem } from '../../utils/bem';

import './EntryItem.scss';

const block = new Bem('entry-item');
const feedBlock = new Bem('entry-item-feed');
const infoBlock = new Bem('entry-item-info');

interface EntryProps {
  entry: Entry;
  isCollapseLong: boolean;
  onView: (id: number) => void;
  onSetRead: (id: number, isRead: boolean) => void;
  onSetFavorite: (id: number, isRead: boolean) => void;
}

function EntryItem({ entry, isCollapseLong, onView, onSetRead, onSetFavorite }: EntryProps): JSX.Element {
  const [isExpanded, setExpanded] = useState(false);
  const itemRef = useRef<HTMLDivElement>(null);
  const titleRef = useRef<HTMLDivElement>(null);

  const onScroll = () => {
    if (!entry.isViewed && itemRef && itemRef.current) {
      const rect = itemRef.current.getBoundingClientRect();
      const height = window.innerHeight / 2;

      let isViewed = false;
      let scrollEnd = false;

      if (window.innerHeight + window.scrollY >= document.body.scrollHeight) {
        scrollEnd = true;
      }

      if ((scrollEnd && rect.top > 0) || (rect.top < height)) {
        isViewed = true;
      }

      if (isViewed) {
        onView(entry.id);
      }
    }
  }

  const onScrollDebounce = debounce(() => {
    onScroll();
  }, 500, false)

  const onMouseMove = () => {
    if (!entry.isViewed && itemRef && itemRef.current) {
      const rect = itemRef.current.getBoundingClientRect();
      const height = rect.top + rect.height;

      if (height < window.innerHeight) {
        onView(entry.id);
      }

      window.removeEventListener('mousemove', onMouseMove);
    }
  }

  const onExpand = () => {
    setExpanded(true);
    onRead();
  };

  const onRead = () => {
    if (!entry.isRead) {
      onSetRead(entry.id, true);
    }
  };

  const onUnread = () => {
    onSetRead(entry.id, false);
  };

  const onClickFavorite = () => {
    onSetFavorite(entry.id, !entry.isFavorite);
  };

  useEffect(() => {
    const element = titleRef.current;

    if (!entry.isViewed) {
      window.addEventListener('scroll', onScrollDebounce);
      window.addEventListener('mousemove', onMouseMove);
    }

    if (element) {
      element.addEventListener('mouseup', onRead);
    }

    return () => {
      window.removeEventListener('scroll', onScrollDebounce);
      window.removeEventListener('mousemove', onMouseMove);

      if (element) {
        element.removeEventListener('mouseup', onRead);
      }
    }
  }, [entry.isViewed]);

  const description = removeSelfLinks(entry.description, entry.link);
  const isCollapsed = isCollapseLong && !isExpanded && isLong(description);
  const blockClass = block.addModifier(isCollapsed ? 'collapsed' : undefined).build();

  return (
    <div ref={itemRef} className={blockClass}>
      <div ref={titleRef} className={block.getElement('title').build()}>
        <a href={entry.link}>{entry.title}</a>
      </div>
      <div className={feedBlock.build()}>
        <div className={feedBlock.getElement('icon').build()} style={{ backgroundImage: `url(${entry.feed.icon})` }} />
        <div className={feedBlock.getElement('title').build()}>{entry.feed.title}</div>
      </div>
      <div className={block.getElement('description').build()} dangerouslySetInnerHTML={{ __html: description }} />
      { isCollapsed && <button className='entry__expandButton' onClick={onExpand} />}
      <div className={infoBlock.build()}>
        <div className={infoBlock.getElement('item').build()}>
          <div className={infoBlock.getElement('icon').build()} onClick={onClickFavorite}>
            {entry.isFavorite ? <FavoriteSelectedIcon /> : <FavoriteIcon /> }
          </div>
        </div>
        { entry.isRead &&
          <div className={infoBlock.getElement('item').build()} onClick={onUnread}>
            <div className={infoBlock.getElement('icon').build()}><ReadIcon /></div>
          </div>
        }
      </div>
    </div>
  );
}

export default EntryItem;
