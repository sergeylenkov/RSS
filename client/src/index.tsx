import React from 'react';
import { createRoot } from 'react-dom/client';
import App from './App';
import { Provider } from 'react-redux';
import store from './store';

import './index.scss';

const container = document.getElementById('root');

if (container) {
  const root = createRoot(container);
  root.render(
    <Provider store={store}>
      <App />
    </Provider>
  );
}
