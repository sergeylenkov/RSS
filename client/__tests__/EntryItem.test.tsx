import React from 'react';
import renderer  from 'react-test-renderer';
import EntryItem from '../src/components/entries/EntryItem';
import { Feed, Entry } from '../src/data';

const action = () => { return };

const feed: Feed = {
  id: 1,
  title: 'Mock feed',
  link: 'https://www.test.com',
  image: '',
  icon: '',
  count: 1
}

const entry: Entry = {
  id: 1,
  title: 'Mock entry',
  description: '',
  link: 'https://www.test.com/1',
  isViewed: false,
  isFavorite: false,
  isRead: false,
  feed: feed,
  feedId: 0
}

test('EntryItem correct rendering', () => {
  const render = renderer.create(<EntryItem entry={entry} isCollapseLong onView={action} onSetRead={action} onSetFavorite={action} />);
  expect(render).toMatchSnapshot();
});