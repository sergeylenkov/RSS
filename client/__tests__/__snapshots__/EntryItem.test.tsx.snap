// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryItem correct rendering 1`] = `
<div
  className="entry-item"
>
  <div
    className="entry-item__title"
  >
    <a
      href="https://www.test.com/1"
    >
      Mock entry
    </a>
  </div>
  <div
    className="entry-item-feed"
  >
    <div
      className="entry-item-feed__icon"
      style={
        Object {
          "backgroundImage": "url()",
        }
      }
    />
    <div
      className="entry-item-feed__title"
    >
      Mock feed
    </div>
  </div>
  <div
    className="entry-item__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
  <div
    className="entry-item-info"
  >
    <div
      className="entry-item-info__item"
    >
      <div
        className="entry-item-info__icon"
        onClick={[Function]}
      >
        <svg
          viewBox="0 0 24 24"
        >
          <path
            d="M17,18L12,15.82L7,18V5H17M17,3H7A2,2 0 0,0 5,5V21L12,18L19,21V5C19,3.89 18.1,3 17,3Z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
